 <h1>Assignment 5.2: Build and Apply Multilayer Perceptron </h1>

Name: Nicolas, Sean Julian S.<br>
Course and Section: CPE019S3 <br>
Instructor: Engr. Roman Richard <br>
Date Performed: March 21, 2024 <br>
Date Submitted: March 26, 2024 <br>

<hr>



In [69]:
pip install ucimlrepo

### Import Libraries

In [70]:
from ucimlrepo import fetch_ucirepo

In [71]:
# importing modules
import tensorflow as tf
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Activation
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

### Get X (features) and Y (target) valeus from data

In [72]:
# fetch dataset
adultdataset = fetch_ucirepo(id=763)

# data (as pandas dataframes)
X = adultdataset.data.features
y = adultdataset.data.targets

### Look at the data to check for missing values and datatypes

In [73]:
X.info()
X.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 338 entries, 0 to 337
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V       338 non-null    float64
 1   H       338 non-null    float64
 2   S       338 non-null    float64
dtypes: float64(3)
memory usage: 8.0 KB


V         H    S
0  0.338157  0.000000  0.0
1  0.320241  0.181818  0.0
2  0.287009  0.272727  0.0
3  0.256284  0.454545  0.0
4  0.262840  0.545455  0.0

In [74]:
y.info()
y.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 338 entries, 0 to 337
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   M       338 non-null    int64
dtypes: int64(1)
memory usage: 2.8 KB


M
0  1
1  1
2  1
3  1
4  1

### Split data and create Perceptron model

In [75]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=40)

model = Sequential([Dense(2000, activation='relu'),Dense(2000, activation='relu'),Dense(2000, activation='relu'),Dense(6, activation='softmax'),])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=13, validation_split=0.2)

Epoch 1/100
17/17 [==============================] - 3s 118ms/step - loss: 1.6427 - accuracy: 0.2593 - val_loss: 1.5354 - val_accuracy: 0.2778
Epoch 2/100
17/17 [==============================] - 2s 114ms/step - loss: 1.4531 - accuracy: 0.2917 - val_loss: 1.4146 - val_accuracy: 0.2037
Epoch 3/100
17/17 [==============================] - 3s 206ms/step - loss: 1.3439 - accuracy: 0.3657 - val_loss: 1.3787 - val_accuracy: 0.3519
Epoch 4/100
17/17 [==============================] - 2s 144ms/step - loss: 1.2588 - accuracy: 0.4213 - val_loss: 1.2923 - val_accuracy: 0.3889
Epoch 5/100
17/17 [==============================] - 3s 160ms/step - loss: 1.1855 - accuracy: 0.4861 - val_loss: 1.2464 - val_accuracy: 0.3519
Epoch 6/100
17/17 [==============================] - 3s 149ms/step - loss: 1.1525 - accuracy: 0.4583 - val_loss: 1.2014 - val_accuracy: 0.3148
Epoch 7/100
17/17 [==============================] - 2s 142ms/step - loss: 1.0652 - accuracy: 0.5046 - val_loss: 1.1712 - val_accuracy: 0.4074

### Evaluate the accuracy of the model

In [77]:
acc = model.evaluate(X_test, y_test, verbose = 1)
print('Test Loss:, Test Accuracy:', acc)

3/3 [==============================] - 0s 20ms/step - loss: 0.8432 - accuracy: 0.7353
Test Loss:, Test Accuracy: [0.8432178497314453, 0.7352941036224365]


### Conclusion:

- The problem of this dataset is to determine the type of landmine that might be enountered in the field. The features are Voltage (coming from the detecor), Height (distance of detector from the soil), and Soil(type of soil). I started from checking for missing data from the dataset and upon checking there was missing data. The next procedure is the creation of the model and the value of dense was arbitrarily decided via multiple trials. In the evaluation part of the assignment, the model showed an accuraacy of 74%. Upon searching for solutions, the only way that I found to increase the accuracy is to increase the epoch (the number of times the data goes through the model) and to choose another dataset with more instances. I guess increasing the number of instances means that output can be more predicted since there are more inputs that can be injected to the model for it to learn.

- Overall, this activity taught me how to make a perceptron model, how to evaluate its performance, and the solutions that can be made in order to increase its accuracy.